src/actions/authActions.js

```diff
@@ -177,20 +177,93 @@ export const loginAction = (
+      if (user?.username) {
         dispatch({ type: SET_USERNAME, payload: user.username });
       }
 
-      if (userState === REGISTERED) {
+      const { address } = wallet;
+      let unlockedWallet = { address };
+
+      /**
+       * if user isn't completely registered then put private key into state to complete registration,
+       * this can happen due offline or web recovery portal onboarding
+       */
+      if (!user?.walletId) {
+        unlockedWallet = { ...unlockedWallet, privateKey: decryptedPrivateKey };
+      }
+
+      // web recovery portal recovery pending, let's navigate accordingly
+      if (getState().wallet.backupStatus.isRecoveryPending) {
+        dispatch({ type: SET_WALLET, payload: unlockedWallet });
+        navigate(NavigationActions.navigate({ routeName: RECOVERY_PORTAL_WALLET_RECOVERY_PENDING }));
+        await smartWalletService.init(
+          decryptedPrivateKey,
+          (event) => dispatch(checkRecoveredSmartWalletStateAction(event)),
+        );
+        dispatch(checkAndFinishSmartWalletRecoveryAction());
+        return;
+      }
+
+      // execute login success callback
+      if (onLoginSuccess) {
+        const rawPrivateKey = decryptedPrivateKey.indexOf('0x') === 0
+          ? decryptedPrivateKey.slice(2)
+          : decryptedPrivateKey;
+        await onLoginSuccess(rawPrivateKey);
+      }
+
+      dispatch({ type: SET_WALLET, payload: unlockedWallet });
+
+      // init smart wallet
+      await dispatch(initOnLoginSmartWalletAccountAction(decryptedPrivateKey));
+
+      const smartWalletAccount = findFirstSmartAccount(accounts);
+
+      // key based wallet migration – switch to smart wallet if key based was active
+      if (getActiveAccountType(accounts) !== ACCOUNT_TYPES.SMART_WALLET && smartWalletAccount) {
+        await dispatch(setActiveAccountAction(smartWalletAccount.id));
+      }
+
+      /**
+       * set Ethereum network as active if we disable feature flag
+       * or end beta testing program while user has set BTC as active network
+       */
+      const revertToDefaultNetwork = !isSupportedBlockchain(blockchainNetwork);
+      if (revertToDefaultNetwork || !blockchainNetwork) {
+        dispatch(setActiveBlockchainNetworkAction(BLOCKCHAIN_NETWORK_TYPES.ETHEREUM));
+      }
+
+      if (isOnline) {
+        // Dispatch action to try and get the latest remote config values...
+        dispatch(loadFeatureFlagsAction());
+
+        // to get exchange supported assets in order to show only supported assets on exchange selectors
+        // and show exchange button on supported asset screen only
+        dispatch(getExchangeSupportedAssetsAction());
+
+        // offline onboarded or very old user that doesn't have any smart wallet, let's create one and migrate safe
+        if (!smartWalletAccount) {
+          // this will import and set Smart Wallet as current active account
+          await dispatch(importSmartWalletAccountsAction(decryptedPrivateKey));
```

Comment:
There are plenty if awaits in our codebase for various dispatch, it's not anti-pattern as even Redux shows examples (i.e. https://redux.js.org/tutorials/essentials/part-5-async-logic#checking-thunk-results-in-components) of ways to do this with `thunk` middleware which we have since the early days and which allows us to have actions as promises.

To answer your question – yes, it's done to wait until action completes.

Reply:
Super